### PR TITLE
🔕 Silent calls

### LINKED ISSUE #786: Support "Silent call"
Server ref: https://github.com/nextcloud/spreed/issues/3950

Basic idea: we need a long press / second call button that allows to start a call without triggering notifications.

### PR DESCRIPTION
**Pull request by Ivansss**
Fix #786

### CODE REVIEW

File: NextcloudTalk/NCChatViewController.m [hunk 1]
```diff
@@ -1072,13 +1083,43 @@ - (void)unreadMessagesButtonPressed:(id)sender
...
+- (void)presentStartCallOptions:(BOOL)audioOnly
+{
+    UIAlertController *optionsActionSheet = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Call options", nil)
+                                                                                message:nil
+                                                                         preferredStyle:UIAlertControllerStyleActionSheet];
+    
+    UIAlertAction *silentSendAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"Call without notification", nil)
+                                                               style:UIAlertActionStyleDefault
+                                                             handler:^void (UIAlertAction *action) {
+        [[CallKitManager sharedInstance] startCall:_room.token withVideoEnabled:NO andDisplayName:_room.displayName silently:YES withAccountId:_room.accountId];
```
Comment (SystemKeeper) [inline NextcloudTalk/NCChatViewController.m]: "Block implicitly retains self"

**Review by SystemKeeper — COMMENTED** [review]
I'm unable to show the action menu when testing on my phone.

Cap is there:
image

GestureRecognizer is added:
image

But no menu on longpress on iPhone (device) and on Mac...

File: NextcloudTalk/NCChatViewController.m (hunk 1 above)
Comment (Ivansss) [reply NextcloudTalk/NCChatViewController.m]: Fixed :)

**Comment by Ivansss** [comment]
Can you try `git submodule update` @SystemKeeper ?
SlackTextViewController was updated in https://github.com/nextcloud/talk-ios/pull/836/commits/6c9ca5cf1f431dc8e5736642ad44b9b858c6cc07

**Comment by SystemKeeper** [comment]
:man_facepalming: Missed it, sorry.

image

Haven't tested on my phone, but works in mac. One minor thing: When starting a call without notifications we loose the activity indicator of the bar button. This might be something we want to look at another time.
Otherwise it looks good.

**Comment by Ivansss** [comment]
Now the activity indicators are shown when starting a call silently.

File: NextcloudTalk/NCChatViewController.m [hunk 2]
```diff
@@ -1072,13 +1083,49 @@ - (void)unreadMessagesButtonPressed:(id)sender
...
+    
+    UIAlertAction *silentSendAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"Call without notification", nil)
+                                                               style:UIAlertActionStyleDefault
+                                                             handler:^void (UIAlertAction *action) {
+        if (audioOnly) {
+            [self->_voiceCallButton showActivityIndicator];
+        } else {
+            [self->_videoCallButton showActivityIndicator];
+        }
+        [[CallKitManager sharedInstance] startCall:self->_room.token withVideoEnabled:NO andDisplayName:self->_room.displayName
```
Comment (SystemKeeper) [inline NextcloudTalk/NCChatViewController.m]: `withVideoEnabled:!audioOnly` ?

File: NextcloudTalk/NCChatViewController.m [hunk 3]
```diff
@@ -1072,13 +1083,49 @@ - (void)unreadMessagesButtonPressed:(id)sender
...
+    }
+}
+
+- (void)presentStartCallOptions:(BOOL)audioOnly
+{
+    UIAlertController *optionsActionSheet = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Call options", nil)
+                                                                                message:nil
+                                                                         preferredStyle:UIAlertControllerStyleActionSheet];
+    
+    UIAlertAction *silentSendAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"Call without notification", nil)
```
Comment (SystemKeeper) [inline NextcloudTalk/NCChatViewController.m]: `silentCallAction` ?

**Review by SystemKeeper — COMMENTED** [review]
Works fine now. Sorry for the confusion.

In general, we have "Send silently" for sending texts without notifications, but have "Call without notification" when starting a call. Should "Send silently" be renamed to "Send without notification" ?

File: NextcloudTalk/NCChatViewController.m (hunk 2 above)
Comment (Ivansss) [reply NextcloudTalk/NCChatViewController.m]: oops 😅

File: NextcloudTalk/NCChatViewController.m (hunk 3 above)
Comment (Ivansss) [reply NextcloudTalk/NCChatViewController.m]: makes more sense yes :D

**Comment by Ivansss** [comment]
> In general, we have "Send silently" for sending texts without notifications, but have "Call without notification" when starting a call. Should "Send silently" be renamed to "Send without notification" ?

Good question! We changed it in the web UI https://github.com/nextcloud/spreed/pull/7454 because "Silent call"was a confusing term. In my opinion "Send silently" next to the 🔕 icon it's good enough but it could be changed to "Send without notification" so it's more consistent.

**Review by SystemKeeper — APPROVED** [review]
👍

**Pull request merged**